url64d634b8f0_agdturner_generic_tgz-pJ8-uk_ac_leeds_ccg_andyt_generic_visualisation_charts_Generic_BarChartJ8/src/uk/ac/leeds/ccg/andyt/generic/math/Generic_double.java

```diff
@@ -0,0 +1,397 @@
+                    System.out.println(
+                            value_BigDecimal + " != " + comparison_BigDecimal);
+                    count_BigInteger = count_BigInteger.add(BigInteger.ONE);
+                    BigDecimal difference = value_BigDecimal.subtract(comparison_BigDecimal);
+                    BigDecimal absDifference = difference.abs();
+                    maxAbsDifference = Generic_BigDecimal.max(maxAbsDifference, absDifference);
+                    if (difference.compareTo(BigDecimal.ZERO) == -1) {
+                        negativeDifferenceSum = negativeDifferenceSum.add(difference);
+                    } else {
+                        positiveDifferenceSum = positiveDifferenceSum.add(difference);
+                    }
+                }
+            }
+            nValues_BigDecimal = Generic_BigDecimal.reciprocal(
+                    increment_BigDecimal, 0, RoundingMode.UNNECESSARY);
+            percentageCorrectValues_BigDecimal = Generic_BigDecimal.divideRoundIfNecessary(
+                    count_BigInteger.multiply(hundred_BigInteger),
+                    nValues_BigDecimal,
+                    5,
+                    RoundingMode.UP);
+            System.out.println(
+                    percentageCorrectValues_BigDecimal.toPlainString() + "% ("
+                    + count_BigInteger.toString() + " out of "
+                    + nValues_BigDecimal.toPlainString()
+                    + ") double values not precisely represented.");
+            System.out.println(
+                    "positiveDifferenceSum " + positiveDifferenceSum.toPlainString());
+            System.out.println(
+                    "negativeDifferenceSum " + negativeDifferenceSum.toPlainString());
+            System.out.println(
+                    "difference "
+                    + positiveDifferenceSum.add(negativeDifferenceSum).toPlainString());
+            System.out.println(
+                    "maxAbsDifference "
+                    + maxAbsDifference.toPlainString());
+            }
+        }
+    }
+
+    public static void print() {
+        System.out.println("Double.MIN_VALUE");
+        System.out.println(toPlainString(Double.MIN_VALUE));
+
+        System.out.println("Double.MIN_NORMAL");
+        System.out.println(toPlainString(Double.MIN_NORMAL));
+
+        System.out.println("Long.MAX_VALUE");
+        System.out.println(Long.MAX_VALUE);
+    }
+
+    /**
+     * Print out the next n a_Random.nextDouble() values that are represented by
+     * a string longer than m
+     * @param a_Random
+     * @param n
+     * @param m
+     */
+    public static void print(
+            Random a_Random,
+            int n,
+            int m) {
+        int counter = 0;
+        BigInteger divisor = new BigInteger("100000000");
+        BigInteger loopCounter = BigInteger.ZERO;
+        while (counter < n) {
+            double value = a_Random.nextDouble();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;a_Random&#39; (index: 0 | type: Random). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;m&#39;, &#39;n&#39;. The same parameter type is validated here for example: url64d634b8f0_agdturner_generic_tgz-pJ8-uk_ac_leeds_ccg_andyt_generic_visualisation_charts_Generic_BarChartJ8/src/uk/ac/leeds/ccg/andyt/generic/utilities/Generic_Collections.java:198. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).